Add Zip file functionality. Fixes #11413

lababidi opened the pull request:
closes #11413

This PR leverages Python's ZipFile functionality to automatically unzip files read into DataFrames using read_csv().

Commit:
Add Zip file functionality. Fixes #11413

Commit:
Add Zip file functionality. Fixes #11413

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -3809,13 +3809,26 @@ def test_decompression(self):
+
+            result = self.read_csv(open(path, 'rb'), compression='zip')
+            tm.assert_frame_equal(result, expected)
+
```
need a test for multiple files in a zip (e.g. assert that the ``ValueError`` is raised)

jreback commented:
as an aside, would be interested in solving issues brought up in https://github.com/pydata/pandas/issues/11666 and partially addressed in #11677 (not merged)

Commit:
Add test to ensure ValueError is thrown when ZIP file contains multiple files.

Commit:
Add test to ensure ValueError is thrown when ZIP file contains multiple files.

lababidi commented:
@jreback I would be happy to continue this to the other `read_*` functions. I like the idea of refactoring out the compression determination and the decompression steps, so that it can be used in all the `read_*` functions. I won't use the PR #11677 because it does not include tests or zip functionality. It also only focuses on `pickles`.

lababidi commented:
@jreback Could you merge this in and I'll take care of the other requests?

jreback commented on pandas/io/parsers.py:
```diff
@@ -61,9 +61,9 @@ class ParserWarning(Warning):
     Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32}
     (Unsupported with engine='python')
-compression : {'gzip', 'bz2', 'infer', None}, default 'infer'
+compression : {'gzip', 'bz2', 'zip', 'infer', None}, default 'infer'
```
can you add an explanation about .zip is only a single file

lababidi committed:
Add parser description warning to handle ZIP files

In the description for the parser, a warning/comment is made that a zip file may only contain one file that needs to be read in. If more than one file is compressed into the ZIP file, a ValueError is thrown.

lababidi committed:
Add parser description warning to handle ZIP files

In the description for the parser, a warning/comment is made that a zip file may only contain one file that needs to be read in. If more than one file is compressed into the ZIP file, a ValueError is thrown.

jreback commented on pandas/parser.pyx:
```diff
@@ -563,6 +563,18 @@ cdef class TextReader:
+                    source = zip_file.open(file_name)
+
+                elif len(zip_names)>1:
+                    raise ValueError('Multiple files found in compressed '
```
maybe just do ``else`` here (e.g. you can have 0 files in an archive?)

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -3809,13 +3809,37 @@ def test_decompression(self):
+
+        with tm.ensure_clean() as path:
+            file_names = ['test_file', 'second_file']
+            tmp = zipfile.ZipFile(path, mode='w')
```
add in another test using ``tmp`` before adding other files (should fail as well)

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -3809,13 +3809,37 @@ def test_decompression(self):
         data = open(self.csv1, 'rb').read()
         expected = self.read_csv(self.csv1)
 
         with tm.ensure_clean() as path:
```
put the zipfile tests in a separate test.

jreback replied:
pls split out the gzip/bz2 into separate tests as well, thanks

jreback replied:
best would be to make a separate testing class for ``TestCompression``

Commit:
Create TestCompression Nose Test Class. Split test_compression into test_gzip, test_bz2, test_zip. Add tests for python and c engines.

Commit:
Create TestCompression Nose Test Class. Split test_compression into test_gzip, test_bz2, test_zip. Add tests for python and c engines.

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge branch 'master' of github.com:pydata/pandas

Conflicts:
	pandas/io/tests/test_parsers.py

lababidi commented:
@jreback could you please merge this before you make anymore changes to the tests.

jreback commented:
you need to rebase then I'll review

Commit:
Merge branch 'master' of github.com:lababidi/pandas

lababidi commented:
@jreback Rebased. Thanks.

jreback commented:
you need to force push. you should only have your commits. see [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-your-changes-to-pandas)

lababidi commented:
@jreback I'm redoing this Pull Request. I will close this one and open a new PR.

jreback commented:
ok, in the future, just push to the same one

itzybitzy commented:
Can the same functionality be added to read_fwf() and other read_* methods?

jreback commented:
#11666 is related
#12688 is where this could be done, e.g. the code is a spread out a bit. welcome for you to take a crack at it @itzybitzy

jreback commented:
IOW, the compression interfaces need to be pulled out a bit from the parser code

itzybitzy commented:
Sorry @jreback, I just don't have the skills to do that (yet). Is there something good for a beginner to work on? Documentation maybe? (Not even sure where the right place to ask this is.)

jreback commented:
http://pandas.pydata.org/pandas-docs/stable/contributing.html

selects label of difficulty novice and you will see lots of issues
